pkg/nsx/services/staticroute/staticroute.go

```diff
@@ -0,0 +1,232 @@
+	staticRouteStore := &StaticRouteStore{}
+	staticRouteStore.Indexer = cache.NewIndexer(keyFunc, cache.Indexers{
+		common.TagScopeStaticRouteCRUID: indexFunc,
+	})
+	staticRouteStore.BindingType = model.StaticRoutesBindingType()
+	service.StaticRouteStore = staticRouteStore
+	service.NSXConfig = commonService.NSXConfig
+	go queryStaticRoute(service, &wg, fatalErrors)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return service, err
+	}
+
+	return service, nil
+}
+
+// assume that staticroute doesn't have the same ipaddress, return true if equal
+func (service *StaticRouteService) compareStaticRoute(oldStaticRoute *model.StaticRoutes, newStaticRoute *model.StaticRoutes) bool {
+	if *oldStaticRoute.Network != *newStaticRoute.Network {
+		return false
+	}
+	oldNextHops := oldStaticRoute.NextHops
+	newNextHops := newStaticRoute.NextHops
+	if len(oldNextHops) != len(newNextHops) {
+		return false
+	}
+	oldHopsSet := sets.NewString()
+	for _, addr := range oldNextHops {
+		oldHopsSet.Insert(*addr.IpAddress)
+	}
+	for _, addr := range newNextHops {
+		if !oldHopsSet.Has(*addr.IpAddress) {
+			return false
+		}
+	}
+	return true
+}
+
+func (service *StaticRouteService) CreateorUpdateStaticRoute(obj *v1alpha1.StaticRoute) error {
+	nsxStaticRoute, err := service.buildStaticRoute(obj)
+	if err != nil {
+		return err
+	}
+
+	existingStaticRoute := service.StaticRouteStore.GetByKey(*nsxStaticRoute.Id)
+	if existingStaticRoute != nil && service.compareStaticRoute(existingStaticRoute, nsxStaticRoute) {
+		return nil
+	}
+	// TODO, update the value from 'default' to actual value
+	err = service.patch("default", "default", "default", nsxStaticRoute)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (service *StaticRouteService) buildStaticRoute(obj *v1alpha1.StaticRoute) (*model.StaticRoutes, error) {
+	sr := &model.StaticRoutes{}
+	sr.Network = &obj.Spec.Network
+	dis := int64(1)
+	for _, nh := range obj.Spec.NextHops {
+		nexthop := model.RouterNexthop{AdminDistance: &dis}
+		nexthop.IpAddress = &nh.IPAddress
+		sr.NextHops = append(sr.NextHops, nexthop)
+	}
+	sr.Id = (*string)(&obj.UID)
+	sr.Tags = service.buildBasicTags(obj)
+	return sr, nil
+}
+
+func (service *StaticRouteService) buildBasicTags(obj *v1alpha1.StaticRoute) []model.Tag {
+	tags := []model.Tag{
+		{
+			Scope: String(common.TagScopeCluster),
+			Tag:   String(service.NSXConfig.Cluster),
+		},
+		{
+			Scope: String(common.TagScopeNamespace),
+			Tag:   String(obj.ObjectMeta.Namespace),
+		},
+		// TODO: get namespace uid
+		{
+			Scope: String(common.TagScopeStaticRouteCRName),
+			Tag:   String(obj.ObjectMeta.Name),
+		},
+		{
+			Scope: String(common.TagScopeStaticRouteCRUID),
+			Tag:   String(string(obj.UID)),
+		},
+	}
+	return tags
+}
+
+func (service *StaticRouteService) patch(orgId string, projectId string, vpcId string, st *model.StaticRoutes) error {
```

Comment:
As discussed, we can use hierarchy mode, then we can refactor those operations. refer [https://github.com/vmware-tanzu/nsx-operator/pull/160/files#diff-58e17097abee8cc07bbcbee5aa217c4e2e98799008f27d875636030d8f7d1f28R103](url)